### PR TITLE
Add documentation for creating large kubernetes clusters

### LINKED ISSUE #10779: Resource management guidance for addons containers
#10653 and #10778 are merged to allocate resource limit (cpu and memory) for all addon containers based on current default configuration for GCE / GKE and AWS: 4 nodes, 30~50 pods per node based on #10335. We should provide scaling guidance and steps for cluster users to update their addon services based on cluster size and workloads. 

cc/ @bgrant0607

### PR DESCRIPTION
**Pull request by saad-ali**
Closes #10779 (adds guidance for addon container scaling)

CC @dchen1107 @satnam6502

### CODE REVIEW

File: docs/cluster-large.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Kubernetes Large Cluster
+
+## Support
+At v1.0, Kubernetes supports clusters up to 100 nodes with 30-50 pods per node and 1-2 container per pod (as defined in the [1.0 roadmap](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/roadmap.md#reliability-and-performance)).
```
Comment (mikedanese) [inline docs/cluster-large.md]: Is there a spec for vertical scale that k8s supports? Does k8s support 32 core nodes? @dchen1107

File: docs/cluster-large.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+## Setup
+
+Normally the number of nodes in a cluster is controlled by the the value `NUM_MINIONS` in the platform-specific `config-default.sh` file (see [GCE's `config-default.sh`](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/config-default.sh), for example).
+
```
Comment (satnam6502) [inline docs/cluster-large.md]: "for example, see..."

File: docs/cluster-large.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  * In GCE, for example, you'll want to increase the quota for:
+    * CPUs
+    * VM instances
+    * Total persistent disk reserved
+    * In-use IP addresses
+    * Firewall Rules
+    * Forwarding rules
+    * Routes
+    * Target pools
+* Gating the setup script so that it brings up new node VMs in smaller batches with waits in between, because some cloud providers limit the number of VMs you can create during a given period.
```
Comment (satnam6502) [inline docs/cluster-large.md]: Perhaps point out where the Quota link is in the UI? Does it have a direct link?

File: docs/cluster-large.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+To avoid running into cluster addon resource issues, when creating a cluster with many nodes, consider the following:
+* Scale memory and CPU limits for each of the following addons, if used, along with the size of cluster (there is one replica of each handling the entire cluster so memory and CPU usage tends to grow proportionally with size/load on cluster):
+  * Heapster ([GCM/GCL backed](cluster/addons/cluster-monitoring/google/heapster-controller.yaml), [InfluxDB backed](cluster/addons/cluster-monitoring/influxdb/heapster-controller.yaml), [GCM/InfluxDB](cluster/addons/cluster-monitoring/googleinfluxdb/heapster-controller-combined.yaml), [standalone](cluster/addons/cluster-monitoring/standalone/heapster-controller.yaml))
+  * [InfluxDB and Grafana](cluster/addons/cluster-monitoring/influxdb/influxdb-grafana-controller.yaml)
+  * [skydns, kube2sky, and dns etcd](cluster/addons/dns/skydns-rc.yaml.in)
+  * [kibana](cluster/addons/fluentd-elasticsearch/kibana-controller.yaml)
```
Comment (satnam6502) [inline docs/cluster-large.md]: Kibana

**Comment by satnam6502** [comment]
Perhaps show a few `kubectl` commands which diagnose a pod being repeatedly killed because it execeeds its resource limits? Just to be concrete?

File: docs/cluster-large.md (hunk 2 above)
Comment (saad-ali) [inline docs/cluster-large.md]: Done.

File: docs/cluster-large.md (hunk 3 above)
Comment (saad-ali) [inline docs/cluster-large.md]: The quota link varies per GCP project. Instead, I linked to the [Resource Quotas help page](https://cloud.google.com/compute/docs/resource-quotas).

File: docs/cluster-large.md (hunk 4 above)
Comment (saad-ali) [inline docs/cluster-large.md]: Done.

File: docs/cluster-large.md (hunk 1 above)
Comment (saad-ali) [inline docs/cluster-large.md]: Mike, I'll let @dchen1107 respond with a definitive answer. But I'm not aware of any spec for vertical scale or any known issues with supporting nodes with large number of cores.

**Comment by saad-ali** [comment]
Thanks for the feedback @satnam6502 
PTAL

**Comment by satnam6502** [comment]
Apart from the mis-aligned headers looks great to me and still LGTM.

**Comment by saad-ali** [comment]
> Apart from the mis-aligned headers...

Sorry I'm probably being dense here, but could you clarify

**Comment by satnam6502** [comment]
```
Events:
  FirstSeen             LastSeen            Count   From                    SubobjectPath               Reason      Message
  Tue, 07 Jul 2015 12:53:51 -0700   Tue, 07 Jul 2015 12:53:51 -0700 1   {scheduler }                        scheduled   Successfully assigned simmemleak-hra99 to kubernetes-minion-tf0f
  Tue, 07 Jul 2015 12:53:51 -0700   Tue, 07 Jul 2015 12:53:51 -0700 1   {kubelet kubernetes-minion-tf0f}    implicitly required container POD   pulled      Pod container image "gcr.io/google_containers/pause:0.8.0" already present on machine
```

Of course, I suspect this is just what kubectl prints out.

**Comment by saad-ali** [comment]
Got it. Yes, that's what kubectl prints. But aligned for clarity.

**Comment by saad-ali** [comment]
@k8s-bot ok to test

File: docs/cluster-large.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  * [skydns, kube2sky, and dns etcd](cluster/addons/dns/skydns-rc.yaml.in)
+  * [Kibana](cluster/addons/fluentd-elasticsearch/kibana-controller.yaml)
+* Scale number of replicas for the following addons, if used, along with the size of cluster (there are multiple replicas of each so increasing replicas should help handle increased load, but, since load per replica also increases slightly, also consider increasing CPU/memory limits):
+  * [elasticsearch](cluster/addons/fluentd-elasticsearch/es-controller.yaml)
+* Increase memory and CPU limits sligthly for each of the following addons, if used, along with the size of cluster (there is one replica per node but CPU/memory usage increases slightly along with cluster load/size as well):
+  * [FluentD with ElasticSearch Plugin](cluster/saltbase/salt/fluentd-es/fluentd-es.yaml)
+  * [FluentD with GCP Plugin](cluster/saltbase/salt/fluentd-gcp/fluentd-gcp.yaml)
+
+
+#### Detecting Resource Starved Containers
```
Comment (erictune) [inline docs/cluster-large.md]: Please move this section to its own document, and link from here, and also link from docs/compute_resources.md.

**Comment by erictune** [comment]
The "detecting resource starved containers" section partially fixes #10108, specifically the request for a section about Setting CPU and Memory Limits

File: docs/cluster-large.md (hunk 5 above)
Comment (saad-ali) [inline docs/cluster-large.md]: Done.

**Comment by saad-ali** [comment]
@erictune Split the `Detecting Resource Starved Containers" section out into a separate doc.
PTAL

**Comment by saad-ali** [comment]
@satnam6502 @erictune Can you PTAL when you get a chance
The 1 failed E2E test looks like a flake

**Comment by satnam6502** [comment]
LGTM. Thanks.

**Comment by erictune** [comment]
try to merge your troubleshooting with the section in compute_resources.md that talks about "OOM Killed" Reason.

**Comment by saad-ali** [comment]
@erictune PTAL

**Comment by bgrant0607** [comment]
See cluster admin doc reorg in #11013.

Will need to be rebased once all of the new munging passes have been merged.

**Comment by saad-ali** [comment]
Ok, I'll rebase if #11013 gets merged first.

**Comment by erictune** [comment]
LGTM

**Pull request merged**